optimum/gptq/quantizer.py

```diff
@@ -0,0 +1,571 @@
+                in the paper ['wikitext2','c4'].
+            tokenizer (`Any`, defaults to `None`):
+                The tokenizer to use in order to prepare the dataset
+            batch_size (`Optional[int]`, *optional*, defaults to `1`):
+                The batch size of the dataset
+            pad_token_id (`Optional[int]`, *optional*, defaults to `None`):
+                The pad token id. Needed to prepare the dataset when `batch_size` > 1.
+
+        Returns:
+            `nn.Module`: The quantized model
+        """
+
+        if not torch.cuda.is_available():
+            raise RuntimeError("No GPU found. A GPU is needed to quantize model.")
+
+        model.eval()
+
+        # For Transformer model
+        has_config = False
+        if hasattr(model, "config"):
+            has_config = True
+        if hasattr(model, "hf_device_map"):
+            # If the model has a device_map, we don't move to model. We have already dispatch the hook that will do the work
+            has_device_map = True
+        device = get_device(model)
+
+        if has_config:
+            use_cache = model.config.use_cache
+            model.config.use_cache = False
+
+        if self.model_seqlen is None:
+            self.model_seqlen = get_seqlen(model)
+
+        # Step 1: Prepare the data
+        if isinstance(dataset, str):
+            dataset, _ = get_dataset(dataset, tokenizer, seqlen=self.model_seqlen)
+        elif isinstance(dataset, list):
+            dataset = [tokenizer(data, return_tensors="pt") for data in dataset]
+        dataset = prepare_dataset(dataset, pad_token_id=pad_token_id, batch_size=batch_size)
+
+        # Step 2: get the input of the 1st block
+        layer_inputs = []
+        layer_outputs = []
+        layer_input_kwargs = []
+
+        class Catcher(nn.Module):
+            """hijack layer's forward pass to cache data"""
+
+            def __init__(self, m):
+                super().__init__()
+                self.module = m
+
+            def forward(self, input=None, **kwargs):
+                # specific to transformers
+                # some models use all key-value arguments in forward pass call
+                if input is None:
+                    if "hidden_states" in kwargs:
+                        input = kwargs["hidden_states"]
+                    else:
+                        raise ValueError("No input value found in the foward pass")
+                layer_inputs.append(input)
+                other_kwargs = {}
+                for k, v in kwargs.items():  # make sure other arguments also be captured
+                    if k not in ["hidden_states"]:
+                        other_kwargs[k] = v
+                layer_input_kwargs.append(other_kwargs)
+                raise ValueError
+
+        # get block_name
+        if self.block_name_to_quantize is None:
+            self.block_name_to_quantize = get_block_name(model)
+
+        # get modules_name that are preceding the first block
```

Comment:
`self.module_name_preceding_first_block = get_preceding_modules(...)` is probably explicit enough, but it could be nice to explain why (just run part of the model up to the given block, correct?)

Reply:
Ofc, the reason is to make sure that the module preceding the first block are on the same device as the first bloc. (We move them to `cuda`) then we run the model. It will stop at the first bloc as we modified the block to raise an Exception.

Reply:
What I meant (and in my next comment) is that this comment is not informative duplicate of `self.module_name_preceding_first_block = get_preceding_modules(...)`), and it can be more interesting to read the "why" we do that.

It's just taste obviously!

Reply:
Thx for the suggestion ! I've removed the comment and added a comment on why we do that